### PR TITLE
 Log config only on rank zero at `torchtune/recipes/full_finetune_distributed.py`

### PR DESCRIPTION
**Pull request by kabir2505**
#### Context
What is the purpose of this PR? Is it to
 - fix a bug
This PR ensures that log_config is only called on rank zero in distributed runs to avoid redundant logging across devices.

Address: #2700 

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
	- Added a rank check to only log config on rank zero

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- I did not change any public API

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -1056,7 +1057,9 @@ def recipe_main(cfg: DictConfig) -> None:
         - Parameters specified in config (see available configs through ``tune ls``)
         - Overwritten by arguments from the command-line
     """
-    config.log_config(recipe_name="FullFinetuneRecipeDistributed", cfg=cfg)
+    if torch.distributed.get_rank() == 0:
```
Comment (SalmanMohammadi) [inline recipes/full_finetune_distributed.py:1060]: Could we move this inside the recipe's setup and use if `self._is_rank_zero`?

**Review by ebsmothers — COMMENTED** [review]
Thanks for the PR @kabir2505! In addition to @SalmanMohammadi's comment, if you want to go a bit further here we can also make the same changes to our other distributed recipes: [full_dpo_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/full_dpo_distributed.py), [lora_dpo_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/lora_dpo_distributed.py), [knowledge_distillation_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/knowledge_distillation_distributed.py), [qat_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/qat_distributed.py), [qat_lora_finetune_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/qat_lora_finetune_distributed.py), [early_exit_finetune_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/dev/early_exit_finetune_distributed.py), [generate_v2_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/dev/generate_v2_distributed.py), [grpo_full_finetune_distributed.py](https://github.com/pytorch/torchtune/blob/main/recipes/dev/grpo_full_finetune_distributed.py), [lora_finetune_distributed_multi_dataset.py](https://github.com/pytorch/torchtune/blob/main/recipes/dev/lora_finetune_distributed_multi_dataset.py)

**Comment by kabir2505** [comment]
Hey @SalmanMohammadi  @ebsmothers — just to confirm, should `config.log_config()` be moved to setup()?

**Comment by joecummings** [comment]
> Hey @SalmanMohammadi @ebsmothers — just to confirm, should `config.log_config()` be moved to setup()?

Correct!

File: recipes/lora_finetune_distributed.py [hunk 2]
```diff
@@ -287,9 +287,9 @@ def setup(self, cfg: DictConfig) -> None:
         """
         if self._is_rank_zero:
             self._metric_logger = config.instantiate(cfg.metric_logger)
-
             # log config with parameter override
```
Comment (joecummings) [inline recipes/lora_finetune_distributed.py:290]: I think this still needs to log config here. The naming is bad, but one logs it to the metric logger (meaning it's saved into the WandB or Tensorboard run), and the other just prints it to disk

**Review by joecummings — COMMENTED** [review]
Couple comments, but looking good

File: recipes/full_finetune_distributed.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 from warnings import warn
 
 import torch
+import torch.distributed
```
[the remark this answers is not in the document]
Comment (kabir2505) [reply recipes/full_finetune_distributed.py:16]: it was for checking rank outside setup earlier, i've fixed it now

File: recipes/qat_distributed.py [hunk 4]
```diff
@@ -1077,7 +1078,16 @@ def recipe_main(cfg: DictConfig) -> None:
         - Parameters specified in config (see available configs through ``tune ls``)
         - Overwritten by arguments from the command-line
     """
-    config.log_config(recipe_name="QATRecipeDistributed", cfg=cfg)
+    if not training.is_distributed():
```
Comment (joecummings) [inline recipes/qat_distributed.py:1081]: Were you running into errors here? Our `tune run` code should catch this.

Comment (kabir2505) [reply recipes/qat_distributed.py:1081]: Hey! Nope, I wasn’t running into any errors. That part was already present in the code—I just added a line/space there, so it shows up as a diff. Nothing was changed functionally.
